File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {number}
 */
var numSubmatrixSumTarget = function(matrix, target) {
  const [m, n] = [matrix.length, matrix[0].length];
  let ans = 0;
  
  for (let i = 0; i < m; i++) {
    for (let j = 1; j < n; j++) {
      matrix[i][j] += matrix[i][j - 1];
    }
  }
  
  for (let k = 0; k < n; k++) {
    for (let j = k; j < n; j++) {
      const map = new Map();
      map.set(0, 1);
      let sum = 0;
      
      for (let i = 0; i < m; i++) {
        sum += matrix[i][j] - (!k ? 0 : matrix[i][k - 1]);
        if (map.has(sum - target)) ans += map.get(sum - target);
        map.set(sum, (map.get(sum) || 0) + 1);
      }
    }
  }
  
  return ans;
};

File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.py
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        m, n, ans = len(matrix), len(matrix[0]), 0
        for row in matrix:
            for j in range(n - 1):
                row[j + 1] += row[j]
        for k in range(n):
            for j in range(k, n):
                d = defaultdict(int)
                d[0], s = 1, 0
                for i in range(m):
                    s += matrix[i][j] - (0 if k == 0 else matrix[i][k - 1])
                    ans += d[s - target]
                    d[s] += 1
        return ans